Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md

```diff
@@ -249,21 +249,19 @@ OSD locations defined in the CRUSH map and topology labels on nodes.
 Refer to the [krbd-options](https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options)
 for more details.
 
-Execute the following steps:
+Execute the following step to enable read affinity for a specific ceph cluster:
 
-* Patch the `rook-ceph-operator-config` configmap using the following
-command.
+* Patch the ceph cluster CR to enable read affinity:
 
 ```console
-kubectl patch cm rook-ceph-operator-config -nrook-ceph -p $'data:\n "CSI_ENABLE_READ_AFFINITY": "true"'
+kubectl patch cephclusters.ceph.rook.io <cluster-name> -n rook-ceph -p '{"spec":{"csi":{"readAffinity":{"enabled": true}}}}'
 ```
 
-* Add topology labels to the Kubernetes nodes. The same labels may be used as mentioned in the
```

Comment:
Mentioning this isn't require anymore?

Reply:
good catch, seems necessary

Reply:
Actually no, That step was needed for external ceph cluster for explicitly specifying the nearest crush location.

We already adopt whatever topology labels are used for internal ceph cluster for read affinity.
We've already documented that advanced configuration here https://github.com/rook/rook/blob/e54f112f04aefdbc6016fae819729c6d8f139818/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-topology

There's no special steps needed for internal ceph cluster.

Reply:
Should we document this at external cluster document then?

Reply:
:+1:  that part will require some elaborate explanation.
I've opened a new doc issue to track this.

https://github.com/rook/rook/issues/13243

Reply:
Thanks!

Reply:
To be clear, the admin owns adding the topology labels to their nodes, so this statement does apply to internal clusters. This is about adding the labels to the nodes, which is different from specifying the labels in the cluster CR which we do automatically. After this is resolved, @BlaineEXE feel free to dismiss my review to merge while i'm out.

Reply:
> To be clear, the admin owns adding the topology labels to their nodes, so this statement does apply to internal clusters. This is about adding the labels to the nodes, which is different from specifying the labels in the cluster CR which we do automatically. After this is resolved, @BlaineEXE feel free to dismiss my review to merge while i'm out.

Both topology labels and labels that need to be added to nodes are the same.

https://github.com/rook/rook/blob/e54f112f04aefdbc6016fae819729c6d8f139818/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-topology

> The topology of the cluster is important in production environments where you want your data spread across failure domains. The topology can be controlled by adding labels to the nodes. When the labels are found on a node at first OSD deployment, Rook will add them to the desired level in the [CRUSH map](https://docs.ceph.com/en/latest/rados/operations/crush-map/).


This needs to be done as part of cluster deployment stage.
So for internal cluster it already exists. Therefore, we don't need to mention it in read affinity section.

Reply:
I think it is still reasonable to add a note that users are expected to have already set up topology labels. It **is** a requirement for this feature, and it is reasonable to note requirements specifically so that we don't hold an assumption that users may be unaware of. Additionally, users can miss things on accident, and a reminder is not harmful. I would suggest changing this to mention that it is a requirement, and link to the section of the deployment doc that provides instructions.

Reply:
> I think it is still reasonable to add a note that users are expected to have already set up topology labels. It is a requirement for this feature, and it is reasonable to note requirements specifically so that we don't hold an assumption that users may be unaware of. Additionally, users can miss things on accident, and a reminder is not harmful. I would suggest changing this to mention that it is a requirement, and link to the section of the deployment doc that provides instructions.

The line existed as part of steps for both internal and external clusters.

For internal ceph cluster, it is not a requirement.

- The [OSD-topology-labels](https://github.com/rook/rook/blob/e54f112f04aefdbc6016fae819729c6d8f139818/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-topology) consists of both custom rook/ceph topology labels and kubernetes topoogy labels. 
- These gets passed onto crush map when OSD is first created.
- Even if the user does not specify them manually, the host and k8s topology which are present by default are still applied.

- We really don't need to bother the user if he hasn't applied such elaborate rook specific rack/row/datacenter labels before hand. If he has, those same labels will anyway be used for read affinity and no extra action is required.
- Changing topology and crush maps at a later point after deployment is too complex and probably not required.


I've opened a new doc issue to track doc changes for external cluster.: https://github.com/rook/rook/issues/13243. 

I've modified docs to add a line linking to the full set of topology labels so user can check which topology labels are being referred to in this section.

Reply:
The topology labels are **not** present by default in all k8s clusters. I still stand by my statement. It is still a requirement that the user set up topology labels before this feature, and users might need reminded of that. Not all users set up topology labels on their K8s clusters. Additionally, the link you mentioned does not document Ceph-only labels. It also includes the inbuilt k8s labels as well, and it does not ask that users fill in more info than they need.

The things that can happen with this documented are:
- users check out the OSD topology labels, see they already set some, and then do nothing
- OR, users check out the OSD topology labels, realize they haven't set them, and then do so, catching a critical mistake they made

The things that can happen with this **removed** from documentation are:
- users check out the OSD topology labels, see they already set some, and then do nothing (same as above)
- users forgot to set topology labels for their clusters, and then they deploy with only the host label, making their topology affinity based on node rather than on zone or whatever else they intend
  - And, because of what you mentioned, "Changing topology and crush maps at a later point after deployment is too complex," it is difficult for users to fix their mistake after having made it

The possible user experiences are worse when this doc is removed. Please leave it in place.